api/v1alpha1/openstackimage_types.go

```diff
@@ -0,0 +1,218 @@
+	Type ImageSourceType `json:"type"`
+
+	// URL describes how to obtain image data by downloading it from a URL. Must be set if Type is 'url'
+	// +unionMember
+	URL *ImageSourceURL `json:"url,omitempty"`
+}
+
+type ImageSourceURL struct {
+	// URL containing image data
+	// +kubebuilder:validation:Format=uri
+	// +kubebuilder:validation:Required
+	URL string `json:"url"`
+
+	// Decompress specifies that the source data must be decompressed with the given compression algorithm before being stored
+	// +optional
+	Decompress *ImageCompression `json:"decompress,omitempty"`
+}
+
+type ImageHash struct {
+	// Algorithm is the hash algorithm used to generate value.
+	// +required
+	Algorithm ImageHashAlgorithm `json:"algorithm"`
+
+	// Value is the hash of the image data using Algorithm.
+	// kubebuilder:validation:MinLength:=1
+	// +required
+	Value string `json:"value"`
+}
+
+// OpenStackImageSpec defines the desired state of OpenStackImage.
+type OpenStackImageSpec struct {
+	// ContainerFormat is the format of the image container.
+	// qcow2 and raw images do not usually have a container, and this can be omitted.
+	// +optional
+	ContainerFormat *ImageContainerFormat `json:"containerFormat,omitempty"`
+
+	// DiskFormat is the format of the disk image.
+	// Normal values are "qcow2", or "raw". Glance may be configured to support others.
+	// +optional
+	DiskFormat *ImageDiskFormat `json:"diskFormat,omitempty"`
+
+	// Protected specifies that the image is protected from deletion.
+	// If not specified, the default is false.
+	// +optional
+	Protected *bool `json:"protected,omitempty"`
+
+	// Tags is a list of tags which will be applied to the image. A tag has a maximum length of 255 characters.
+	// +optional
+	Tags []ImageTag `json:"tags,omitempty"`
+
+	// Visibility of the image
+	// +optional
+	Visibility *ImageVisibility `json:"visibility,omitempty"`
+
+	// AdditionalProperties allows arbitrary glance properties to be set. These will be merged
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	AdditionalProperties []ImageAdditionalProperty `json:"additionalProperties,omitempty"`
+}
+
+// OpenStackImageStatus defines the observed state of OpenStackImage.
+type OpenStackImageStatus struct {
+	// ID is the UUID of the glance image
+	ID *string `json:"id,omitempty"`
+
+	// Hash is the hash of the image data calculated by glance
+	Hash *ImageHash `json:"hash,omitempty"`
+
+	// SizeB is the size of the image data, in bytes
+	SizeB *int64 `json:"sizeB,omitempty"`
+
+	// VirtualSizeB is the size of the disk the image data represents, in bytes
+	VirtualSizeB *int64 `json:"virtualSizeB,omitempty"`
+
+	// Conditions represents the
+	// Known .status.conditions.type are: "Available", "Progressing", and "Status"
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,1,rep,name=conditions"`
```

Comment:
Idiomatically the conditions should be the first item in the list. Don't need the protobuf marker on this line either

Reply:
I'd wondered about the protobuf thing! Also patchMergeKey and patchStrategy.

I'll take them out, but it's probably worth noting that I copied them from the docstring of the Condition struct: https://github.com/kubernetes/kubernetes/blob/7060e485694e82698c050a6a6c6ae2af20cf5e57/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1517-L1528